crates/dbs-miniball/README.md

```diff
@@ -1,21 +1,528 @@
+[`dbs-legacy-devices`](https://github.com/openanolis/dragonball-sandbox/tree/main/crates/dbs-legacy-devices) and
+[`dbs-utils`](https://crates.io/crates/dbs-utils).
+
+### Architecture
+
+The Miniball is composed of 2 major crates packaged in a binary
+application: the `vmm` and a oneshot CLI that invokes it. The `vmm` crate
+exports an object (`struct Vmm`) that encapsulates all the
+functionality-providing `rust-vmm` crates, as dependencies.
+
+Users of the Miniball can replace the builtin command line parser with
+frontends of their own, as long as they are passing a `VMMConfig` object to
+the `vmm` crate (object that will store specialized `*Config` objects, e.g.
+`struct MemoryConfig` for guest memory configurations, etc.). Alternatively,
+users can disregard the `vmm` crate altogether and build their own, referring
+to it only for guidelines.
+
+![Overview](docs/overview.png)
+
+## Steps to running a guest
+
+Currently, the only available devices are the serial console, and a single
+block device, that is used as the root filesystem. Therefore, the Miniball
+can boot a simple Linux kernel either with a baked-in `initramfs` (no
+block device configured) or with a separate rootfs.
+`*Config` structs passed by the UI are cached in a `VmmConfig` struct until
+`VMM::try_from()`, when the VMM is instantiated and the configurations are
+turned into live objects.
+
+The steps to running a guest, presuming the configurations are parsed and
+stored, are as follows:
+
+1. Set up KVM. This is done through `kvm-ioctls`. It creates the KVM virtual
+   machine in the host kernel.
+
+```rust
+// src/vmm/src/lib.rs
+
+let kvm = Kvm::new().map_err(Error::KvmIoctl)?;
+
+// Check that the KVM on the host is supported.
+let kvm_api_ver = kvm.get_api_version();
+if kvm_api_ver != KVM_API_VERSION as i32 {
+    return Err(Error::KvmApiVersion(kvm_api_ver));
+}
+Vmm::check_kvm_capabilities(&kvm)?;
+```
+
+2. Configure guest memory. This is done through the `vm-memory` crate. This
+   creates and registers the guest memory with KVM.
+   1. Requirements: KVM set up
+   2. Inputs
+       1. guest memory size
+
+```rust
+// src/vmm/src/lib.rs
+
+let guest_memory = Vmm::create_guest_memory(&config.memory_config)?;
+let address_allocator = Vmm::create_address_allocator(&config.memory_config)?;
+let device_mgr = Arc::new(Mutex::new(IoManager::new()));
+...
+let vm = KvmVm::new(
+    &kvm,
+    vm_config,
+    &guest_memory,
+    wrapped_exit_handler.clone(),
+    device_mgr.clone(),
+)?;
+```
+
+3. Create event manager for device events. This is done through `event-manager`.
+
+```rust
+// src/vmm/src/lib.rs
+
+let mut event_manager = EventManager::<Arc<Mutex<dyn MutEventSubscriber + Send>>>::new()
+.map_err(Error::EventManager)?;
+event_manager.add_subscriber(wrapped_exit_handler.0.clone());
+```
+
+4. Configure the vCPUs. This is done through `vm-vcpu` crate, which for now is a
```

Comment:
Actually, lots of functions here could be accomplished by `dbs-boot`, `dbs-arch`